piet-coregraphics/src/text.rs

```diff
@@ -0,0 +1,433 @@
+//! Text related stuff for the coregraphics backend
+
+use std::marker::PhantomData;
+
+use core_foundation_sys::base::CFRange;
+use core_graphics::base::CGFloat;
+use core_graphics::context::CGContextRef;
+use core_graphics::geometry::{CGPoint, CGRect, CGSize};
+use core_graphics::path::CGPath;
+use core_text::font::{self, CTFont};
+
+use piet::kurbo::{Point, Size};
+use piet::{
+    Error, Font, FontBuilder, HitTestMetrics, HitTestPoint, HitTestTextPosition, LineMetric, Text,
+    TextLayout, TextLayoutBuilder,
+};
+
+use crate::ct_helpers::{AttributedString, Frame, Framesetter, Line};
+
+// inner is an nsfont.
+#[derive(Debug, Clone)]
+pub struct CoreGraphicsFont(CTFont);
+
+pub struct CoreGraphicsFontBuilder(Option<CTFont>);
+
+#[derive(Clone)]
+pub struct CoreGraphicsTextLayout {
+    string: String,
+    attr_string: AttributedString,
+    framesetter: Framesetter,
+    pub(crate) frame: Option<Frame>,
+    // distance from the top of the frame to the baseline of each line
+    pub(crate) line_y_positions: Vec<f64>,
+    /// offsets in utf8 of lines
+    line_offsets: Vec<usize>,
+    pub(crate) frame_size: Size,
+    width_constraint: f64,
+}
+
+pub struct CoreGraphicsTextLayoutBuilder(CoreGraphicsTextLayout);
+
+pub struct CoreGraphicsText<'a>(PhantomData<&'a ()>);
```

Comment:
This seems strange. Is there a better type (the &self when `Text` gets created) that would document the expected lifetime nesting?

Reply:
This is the same as what we do in cairo, currently.

we could pass this in a reference to the creating context, but I'm not sure that's any more meaningful?

Reply:
I'm fine leaving it as it is, it's basically documentation.